# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/NativeObject-inl.h
# Commit: 211c4f80a286
# Full Hash: 211c4f80a2862e257dd0610274e5021243920404
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-06 09:33:16
# Regressor Bug: 1689413
# File Overlap Count: 1
# Description:
#   Bug 1689413 part 3 - Simplify NativeObject::removeProperty a bit. r=jonco
#   
#   The `BaseShape::get` call is redundant. Reusing the BaseShape from the previous
#   last-property is equivalent and infallible, and lets us simplify the code a bit.
#   
# ==============================================================================

diff -r 6f7514b0a657 -r 211c4f80a286 js/src/vm/NativeObject-inl.h
--- a/js/src/vm/NativeObject-inl.h	Fri Mar 05 19:10:05 2021 +0000
+++ b/js/src/vm/NativeObject-inl.h	Fri Mar 05 19:10:05 2021 +0000
@@ -57,16 +57,18 @@
 }
 
 inline bool NativeObject::canRemoveLastProperty() {
-  /*
-   * Check that the information about the object stored in the last
-   * property's base shape is consistent with that stored in the previous
-   * shape. If not consistent, then the last property cannot be removed as it
-   * will induce a change in the object itself, and the object must be
-   * converted to dictionary mode instead. See BaseShape comment in jsscope.h
-   */
+  // Check that the information about the object stored in the last property's
+  // Shape is consistent with that stored in the previous shape. If not
+  // consistent, then the last property cannot be removed as it will induce a
+  // change in the object itself, and the object must be converted to dictionary
+  // mode instead.
   MOZ_ASSERT(!inDictionaryMode());
   Shape* previous = lastProperty()->previous().get();
-  return previous->objectFlags() == lastProperty()->objectFlags();
+  if (previous->objectFlags() != lastProperty()->objectFlags()) {
+    return false;
+  }
+  MOZ_ASSERT(lastProperty()->base() == previous->base());
+  return true;
 }
 
 inline void NativeObject::initDenseElementHole(uint32_t index) {